src/xrpld/overlay/Slot.h

```diff
@@ -593,244 +330,181 @@ class Slots final
-    {
-        auto const& it = slots_.find(validator);
-        if (it != slots_.end())
-            return it->second.state_ == state;
-        return false;
-    }
-
-    /** Get selected peers */
-    std::set<id_t>
-    getSelected(PublicKey const& validator)
-    {
-        auto const& it = slots_.find(validator);
-        if (it != slots_.end())
-            return it->second.getSelected();
-        return {};
-    }
-
-    /** Get peers info. Return map of peer's state, count, and squelch
-     * expiration milliseconds.
-     */
-    std::unordered_map<
-        typename Peer::id_t,
-        std::tuple<PeerState, uint16_t, uint32_t, std::uint32_t>>
-    getPeers(PublicKey const& validator)
-    {
-        auto const& it = slots_.find(validator);
-        if (it != slots_.end())
-            return it->second.getPeers();
-        return {};
-    }
-
-    /** Get Slot's state */
-    std::optional<SlotState>
-    getState(PublicKey const& validator)
-    {
-        auto const& it = slots_.find(validator);
-        if (it != slots_.end())
-            return it->second.getState();
-        return {};
-    }
-
     /** Called when a peer is deleted. If the peer was selected to be the
      * source of messages from the validator then squelched peers have to be
      * unsquelched.
      * @param id Peer's id
      * @param erase If true then erase the peer
      */
     void
-    deletePeer(id_t id, bool erase);
+    deletePeer(Peer::id_t id, bool erase);
+
+    /** Called to register that a given validator was squelched for a given
+     * peer. It is expected that this method is called by SquelchHandler.
+     *
+     * @param validatorKey Validator public key
+     * @param peerID peer ID
+     */
+    void
+    squelchValidator(PublicKey const& validatorKey, Peer::id_t peerID);
+
+    void
+    onWrite(beast::PropertyStream::Map& stream) const;
 
 private:
     /** Add message/peer if have not seen this message
      * from the peer. A message is aged after IDLED seconds.
      * Return true if added */
     bool
-    addPeerMessage(uint256 const& key, id_t id);
+    addPeerMessage(uint256 const& key, Peer::id_t id);
+
+    /**
+     * Updates the last message sent from a validator.
+     * @param validator The validator public key
+     * @param peer The peer ID sending the message
+     * @return true if the validator was updated, false otherwise
+     */
+    std::optional<PublicKey>
+    updateConsideredValidator(PublicKey const& validator, Peer::id_t peer);
+
+    /** Remove all validators that have become invalid due to selection
+     * criteria
+     * @return zero or more validators that have been removed.
+     */
+    std::vector<PublicKey>
+    cleanConsideredValidators();
+
+    /** Checks whether a given validator is squelched.
+     * @param validatorKey Validator public key
+     * @return true if a given validator was squelched
+     */
+    bool
+    validatorSquelched(PublicKey const& validatorKey);
```

Comment:
This and next one should be split into two functions. One is public const to return the state and another expire and return the state. Otherwise they could be used to check for the state and unintentionally expire. Perhaps something like `expireAndIsValidatorSquelched`.

Reply:
This is intended behaviour. If a squelch for a validator or a peer has expired, it should no longer be squelched, thus expiring old entries is always a desired behaviour.  Separating this logic into two separate methods would mean that a validator or a peer whose squelch has expired would wrongfully be marked as squelched, where in reality they're expired.